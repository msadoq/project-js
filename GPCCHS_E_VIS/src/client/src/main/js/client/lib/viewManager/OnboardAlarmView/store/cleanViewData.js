/* eslint-disable no-continue, "DV6 TBC_CNES Perf. requires 'for', 'continue' avoid complexity" */
import _get from 'lodash/get';
import _isEqual from 'lodash/isEqual';
import _last from 'lodash/last';
import _pick from 'lodash/pick';
import _findIndex from 'lodash/findIndex';
import _findLastIndex from 'lodash/findLastIndex';
import _pickBy from 'lodash/pickBy';
import _union from 'lodash/union';
import _map from 'lodash/map';
import getLogger from '../../../common/logManager';
import * as constants from '../../../constants';

const logger = getLogger('view:OnBoardAlarmView:cleanViewData');

/* ************************************************
 * Clean viewData for current viewData
 * @param currentState view data State
 * @param oldViewFromMap current view definition
 * @param newViewFromMap current view definition
 * @param oldIntervals expected intervals for all entry points
 * @param newIntervals expected intervals for all entry points
 * @return cleaned state for current view
/** *********************************************** */
export default function cleanCurrentViewData(
  currentState,
  oldViewFromMap,
  newViewFromMap,
  oldIntervals,
  newIntervals
) {
  // Check if viewMap has changed
  if (_isEqual(newViewFromMap, oldViewFromMap) && _isEqual(oldIntervals, newIntervals)) {
    return currentState;
  }
  // new visible view
  if (!oldViewFromMap || !currentState) {
    return currentState;
  }
  // invisible view
  if (!newViewFromMap) {
    return null;
  }
  // entry point updates
  const oldEntryPoints = oldViewFromMap.entryPoints;
  const newEntryPoints = newViewFromMap.entryPoints;
  const epNames = Object.keys(oldEntryPoints);
  // only one entry point is displayed in a OnBoard alarm view
  if (epNames.length !== 1) {
    return currentState;
  }
  const epName = epNames[0];
  const oldEp = oldEntryPoints[epName];
  const newEp = newEntryPoints[epName];

  // removed entry point if invalid
  // EP definition modified: remove entry point from viewData
  if (isInvalidEntryPoint(oldEp, newEp)) {
    return null;
  }
  // Case of point already in error
  if (newEp.error) {
    return currentState;
  }
  let newState = currentState;
  // update on expected interval
  // If EP is valid, old and new tbdId are the same
  // Consider new localId to take into account offset modification
  const oldInterval = _get(oldIntervals, [oldEp.tbdId, oldEp.localId, 'expectedInterval']);
  const newInterval = _get(newIntervals, [oldEp.tbdId, newEp.localId, 'expectedInterval']);
  if (!newInterval || oldEp.localId !== newEp.localId) {
    return null;
  } else if (oldInterval &&
    (oldInterval[0] !== newInterval[0] || oldInterval[1] !== newInterval[1])) {
    const lower = newInterval[0] + newEp.offset;
    const upper = newInterval[1] + newEp.offset;
    newState = removeViewDataOutOfBounds(newState, lower, upper);
  }

  return newState;
}

function isInvalidEntryPoint(oldEp, newEp) {
  if (!newEp || (newEp.error && newEp.error !== oldEp.error)
    || oldEp.tbdId !== newEp.tbdId) {
    return true;
  }
  return false;
}

/**
 * Get all alarms requiring an ack
 * @param  {object} viewData
 * @return {array}  array of filtered indexes
 */
function getRequireAckIndexes(viewData) {
  const requireAckAlarms = _pickBy(viewData.lines, alarm => (
    alarm.ackState === constants.OBA_ALARM_ACKSTATE_REQUIREACK
  ));

  return _map(requireAckAlarms, viewDataAlarm => viewDataAlarm.oid);
}

/**
 * (Immutable) remove data out of [lower, upper] time range from viewData
 *
 * @param  {Object} viewData
 * @param  {number} lower    lower timestamp
 * @param  {number} upper    upper timestamp
 *
 * @return {Object}          The computed viewdata
 */
function removeViewDataOutOfBounds(viewData, lower, upper) {
  if (lower > upper) {
    logger.warn('Received invalid bounds');
    return null;
  }

  // --- Keep everything --- //

  // state contains no data
  if (!viewData || !viewData.indexes || !viewData.indexes.length) {
    return viewData;
  }

  // All points of entryPoint are in visuWindow
  const firstTimestamp = viewData.lines[viewData.indexes[0]].timestamp;
  const lastTimestamp = viewData.lines[_last(viewData.indexes)].timestamp;
  if (firstTimestamp >= lower && lastTimestamp <= upper) {
    return viewData;
  }

  // --- Drop everything but alarms requiring ack --- //

  if (firstTimestamp > upper || lastTimestamp < lower) {
    const newIndexes = getRequireAckIndexes(viewData);
    const newLines = _pick(viewData.lines, newIndexes);

    return {
      lines: newLines,
      indexes: newIndexes,
    };
  }

  // --- Keep some --- //
  const iLower = _findIndex(viewData.indexes, searchOid => (
    viewData.lines[searchOid].timestamp >= lower
  ));
  let iUpper = _findLastIndex(viewData.indexes, searchOid => (
    viewData.lines[searchOid].timestamp <= upper
  ));
  iUpper = (iUpper === -1) ? viewData.lines.length - 1 : iUpper;

  let newIndexes = viewData.indexes.slice(iLower, iUpper + 1);
  newIndexes = _union(newIndexes, getRequireAckIndexes(viewData)).sort((a, b) => (
    viewData.lines[a].timestamp - viewData.lines[b].timestamp
  ));
  const newLines = _pick(viewData.lines, newIndexes);

  return {
    lines: newLines,
    indexes: newIndexes,
  };
}
