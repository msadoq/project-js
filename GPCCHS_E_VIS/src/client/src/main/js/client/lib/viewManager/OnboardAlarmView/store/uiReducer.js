import _ from 'lodash/fp';

export default _.always({});
