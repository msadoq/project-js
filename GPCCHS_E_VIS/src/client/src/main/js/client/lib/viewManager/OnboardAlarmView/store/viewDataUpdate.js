/* eslint-disable no-continue, "DV6 TBC_CNES Perf. requires 'for', 'continue' avoid complexity" */
import _ from 'lodash/fp';
import _findIndex from 'lodash/findIndex';
import _cloneDeep from 'lodash/cloneDeep';
import _concat from 'lodash/concat';
import _get from 'lodash/get';
import _each from 'lodash/each';
// import { applyFilters } from '../../commonData/applyFilters';
import { convertData } from '../../commonData/convertData';
import * as constants from '../../../constants';


/**
 * Debuging function. To check if the indexes are well sorted.
 */
// function logTheTimestamps(viewData) {
//   console.log(viewData.indexes.map(oid => (
//     viewData.lines[oid].timestamp
//   )));
// }

/**
 * Add payloads in plot view data state
 *
 * @param  {Object} state    data state of current view
 * @param  {string} viewId   current view ID
 * @param  {Object} payloads data to add in state per EP name
 *
 * @return {Object}          updated state
 */
export function viewRangeAdd(state = {}, viewId, payloads) {
  // get EP names
  const epNames = Object.keys(payloads || {});
  if (epNames.length !== 1) {
    // no data
    return state;
  }

  // Loop on payloads to update state
  // lines: contains all fields indexed by oid { [oid]: { values }}
  // indexes: contains ordered oids [oid1, oid2, ...]
  let newState = _cloneDeep(state);
  if (!newState.indexes) {
    newState = { lines: {}, indexes: [] };
  }

  // loop on EP name to add payload sorted by oid in EP table
  const epName = epNames[0];

  // Update of EP data
  newState.lines = Object.assign({}, newState.lines, payloads[epName]);
  const oids = Object.keys(payloads[epName]);

  let lastIndex = -1;
  let lastTime;
  // loop on payload oids
  for (let oidIndex = 0; oidIndex < oids.length; oidIndex += 1) {
    const insertOid = oids[oidIndex];
    const payload = payloads[epName][insertOid];
    const time = payload.timestamp;
    // Optimisation when payload is sorted by ascending time
    if (lastIndex === -1 && lastTime && lastTime <= time) {
      newState.indexes.push(insertOid);
    } else {
      // search where to insert payload
      let insertIndex = -1;
      if (newState.indexes.length) {
        insertIndex = _findIndex(newState.indexes, searchOid => (
          newState.lines[searchOid].timestamp >= time
        ));
      }
      lastIndex = insertIndex;
      if (insertIndex === -1) {
        newState.indexes.push(insertOid);
      } else if (newState.indexes[insertIndex] !== insertOid) {
        newState.indexes = _concat(
          newState.indexes.slice(0, insertIndex),
          insertOid,
          newState.indexes.slice(insertIndex));
      } else {
        // Data is already present in table lines, no need to add it
      }
    }
    lastTime = time;
  }

  // logTheTimestamps(newState);
  return newState;
}

/* ************************************
 * Select payload to add for current view
 * @param: current view data map
 * @param: intervals for all entry Points
 * @param: received data
 * @return: payloads to use per EP (unique)
/* *********************************** */
export function selectDataPerView(currentViewMap, intervalMap, payload, visuWindow) {
  let epSubState = {};
  if (currentViewMap) {
    const epNames = Object.keys(currentViewMap.entryPoints);
    // Only one entry point per onBoard alarm view
    if (epNames.length !== 1) {
      return {};
    }
    const epName = epNames[0];
    const ep = currentViewMap.entryPoints[epName];
    // No payload for this tbd  Id
    if (!payload[ep.tbdId]) {
      return {};
    }
    epSubState = selectEpData(payload[ep.tbdId], ep, epName, intervalMap, visuWindow);
  }
  return epSubState;
}


/* ************************************
 * Select payload to add for current entry Point
 * @param: payload of current entry point
 * @param: entry point definition
 * @param: entry point name
 * @param: intervals for all entry Points
 * @return: ep payload to be used
/* *********************************** */
export function selectEpData(tbdIdPayload, ep, epName, intervalMap) {
  // get expected interval
  const expectedInterval = _get(intervalMap, [ep.tbdId, ep.localId, 'expectedInterval']);
  // case of error when isuWindow duration is too long
  if (!expectedInterval) {
    return {};
  }
  const lower = expectedInterval[0];
  const upper = expectedInterval[1];
  const epSubState = { [epName]: {} };

  // Loop on payload timestamps
  _each(tbdIdPayload, (currentValue, i) => {
    const onBoardAlarm = currentValue.onBoardAlarm;
    const oid = currentValue.oid;
    if (!oid || !onBoardAlarm) {
      return;
    }

    const offset = ep.offset || 0;
    const timestamp = (onBoardAlarm.onBoardDate.value || Number(i)) + offset;
    // TODO: needs to determine on which filters have top be applied
    // // check value verify filters
    // if (!applyFilters(currentValue, ep.filters)) {
    //   continue;
    // }

    // Compute acknowledgement State
    let ackState = constants.OBA_ALARM_ACKSTATE_NOACK;
    if (currentValue.ackRequest) {
      ackState = constants.OBA_ALARM_ACKSTATE_REQUIREACK;
      if (currentValue.ackRequest && currentValue.ackRequest.ack) {
        ackState = constants.OBA_ALARM_ACKSTATE_ACQUITED;
      }
    }

    if (ep.mode === constants.OBA_ALARM_MODE_TOACKNOWLEDGE) {
      if (ackState !== constants.OBA_ALARM_ACKSTATE_REQUIREACK) {
        return;
      }
    }

    // Filter values out of interval but keep "REQUIREACK" Alarms
    const isOutOfTimeRange = timestamp < lower || timestamp > upper;
    if (isOutOfTimeRange && ackState !== constants.OBA_ALARM_ACKSTATE_REQUIREACK) {
      return;
    }

    const parameters = onBoardAlarm.parameter || [];
    const valueToInsert = {
      oid,
      timestamp,
      ackState,
      collapsed: true,
      satellite: convertData(currentValue.satellite),
      telemetryType: convertData(currentValue.telemetryType),
      onBoardDate: convertData(onBoardAlarm.onBoardDate),
      alarmType: convertData(onBoardAlarm.alarmLevel),
      RIDId: convertData(onBoardAlarm.reportId),
      RIDName: convertData(onBoardAlarm.reportName),
      reportType: convertData(onBoardAlarm.eventType),
      parameters: _.map(_.mapValues(convertData), parameters),
    };
    epSubState[epName][oid] = valueToInsert;
  });

  // if no data, return empty state
  if (!Object.keys(epSubState[epName]).length) {
    return {};
  }
  return epSubState;
}
