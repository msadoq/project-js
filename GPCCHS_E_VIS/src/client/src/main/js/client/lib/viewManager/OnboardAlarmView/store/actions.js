import _ from 'lodash/fp';
import { v4 } from 'uuid';
import simple from '../../../store/helpers/simpleActionCreator';
import * as types from '../../../store/types';
import { getCurrentVisuWindow } from '../../../store/selectors/timebars';
import { openInCurrentWindow as openModalInCurrentWindow } from '../../../store/actions/modals';

export const updateAlarmDomain = simple(types.WS_VIEW_UPDATE_ALARMDOMAIN, 'viewId', 'domainName');
export const updateAlarmTimeline = simple(types.WS_VIEW_UPDATE_ALARMTIMELINE, 'viewId', 'timelineName');
export const updateAlarmMode = (viewId, mode) => (dispatch, getState) => {
  const visuWindow = getCurrentVisuWindow(getState());
  dispatch({ type: types.WS_VIEW_UPDATE_ALARMMODE, payload: { viewId, mode, visuWindow } });
};

export const openAckModal = (viewId, alarmsTimestamps) => {
  const ackId = v4();
  return openModalInCurrentWindow({
    type: 'obaAck',
    title: 'On Board Alarm',
    ackId,
    viewId,
    alarmsTimestamps,
  });
};

export const sendAlarmAck = (alarms, comment) => ({
  type: types.WS_VIEW_OBA_ALARM_ACK,
  payload: {
    alarmsTimestamps: _.isArray(alarms) ? alarms : [alarms],
    comment,
  },
});
