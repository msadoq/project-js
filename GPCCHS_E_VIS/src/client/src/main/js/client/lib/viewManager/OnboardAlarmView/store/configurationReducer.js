import _ from 'lodash/fp';
import { createSelector } from 'reselect';

import * as types from 'store/types';
import { getConfigurationByViewId } from 'viewManager';

export default (stateConf = {}, action) => {
  switch (action.type) {
    case types.WS_VIEW_RELOAD:
    case types.WS_VIEW_OPENED:
    case types.WS_PAGE_OPENED:
    case types.WS_WORKSPACE_OPENED:
    case types.WS_VIEW_ADD_BLANK: {
      return action.payload.view.configuration;
    }
    case types.WS_VIEW_UPDATE_ALARMDOMAIN:
      return _.set('entryPoints[0].connectedData.domain', action.payload.domainName, stateConf);
    case types.WS_VIEW_UPDATE_ALARMTIMELINE:
      return _.set('entryPoints[0].connectedData.timeline', action.payload.timelineName, stateConf);
    case types.WS_VIEW_UPDATE_ALARMMODE: {
      return _.set('entryPoints[0].connectedData.mode', action.payload.mode, stateConf);
    }
    default:
      return stateConf;
  }
};

export const getAlarmMode = createSelector(
  getConfigurationByViewId,
  _.get('entryPoints[0].connectedData.mode')
);

export const getAlarmDomain = createSelector(
  getConfigurationByViewId,
  _.get('entryPoints[0].connectedData.domain')
);

export const getAlarmTimeline = createSelector(
  getConfigurationByViewId,
  _.get('entryPoints[0].connectedData.timeline')
);
