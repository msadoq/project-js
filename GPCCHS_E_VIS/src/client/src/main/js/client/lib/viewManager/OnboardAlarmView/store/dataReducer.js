import _ from 'lodash/fp';
import { createSelector } from 'reselect';
import _omit from 'lodash/omit';

import cleanCurrentViewData from './cleanViewData';
import { viewRangeAdd, selectDataPerView } from './viewDataUpdate';

import * as types from '../../../store/types';
import * as constants from '../../constants';

const initialSubState = {
  lines: {},
  indexes: [],
  ackStatus: {},
};

/* eslint-disable complexity, "DV6 TBC_CNES Redux reducers should be implemented as switch case" */
export default function onBoardAlarmViewData(state = initialSubState, action) {
  switch (action.type) {
    case types.DATA_REMOVE_ALL_VIEWDATA:
    case types.HSC_CLOSE_WORKSPACE:
      return {};
    case types.WS_VIEW_RELOAD:
    case types.WS_VIEW_OPENED:
    case types.WS_VIEW_ADD_BLANK:
      if (action.payload.view.type !== constants.VM_VIEW_ONBOARDALARM) {
        return state;
      }
      return { ...state, [action.payload.view.uuid]: initialSubState };
    case types.WS_PAGE_OPENED:
    case types.WS_WORKSPACE_OPENED:
      {
        const { views } = action.payload;
        if (!views) {
          return state;
        }
        const newState = {};
        views.forEach((view) => {
          if (view.type !== constants.VM_VIEW_ONBOARDALARM) {
            return;
          }
          newState[view.uuid] = initialSubState;
        });
        return { ...state, ...newState };
      }
    case types.WS_VIEW_CLOSE: {
      const { viewId } = action.payload;
      if (state[viewId]) {
        return _omit(state, viewId);
      }
      return state;
    }
    case types.WS_PAGE_CLOSE: {
      const { viewIds } = action.payload;
      if (!viewIds.length) {
        return state;
      }
      let newState = state;
      viewIds.forEach((viewId) => {
        if (state[viewId]) {
          newState = _omit(newState, viewId);
        }
      });
      return newState;
    }
    case types.WS_MODAL_OPEN: {
      const { type, viewId, ackId, alarmsOids } = action.payload.props;
      if (type === 'obaAck') {
        return _.set([viewId, 'ackStatus', ackId], {
          acknowledging: false,
          alarmsOids: alarmsOids.map(oid => ({
            oid,
            acknowledged: false,
            ackError: null,
          })),
        }, state);
      }
      return state;
    }
    case types.WS_MODAL_CLOSED: {
      const { type, viewId, ackId } = action.payload.props;
      if (type === 'obaAck') {
        return _.unset([viewId, 'ackStatus', ackId], state);
      }
      return state;
    }
    case types.WS_VIEW_OBA_ALARM_ACK: {
      const { viewId, ackId } = action.payload;
      const newState = _.set([viewId, 'ackStatus', ackId, 'acknowledging'], true, state);
      return newState;
    }
    case types.WS_VIEW_OBA_ALARM_ACK_SUCCESS: {
      const { viewId, ackId, oid } = action.payload;
      if (!state[viewId].ackStatus[ackId]) {
        return state;
      }
      const { alarmsOids } = state[viewId].ackStatus[ackId];
      const iOid = _.findIndex(_.propEq('oid', oid), alarmsOids);
      return _.set([viewId, 'ackStatus', ackId, 'alarmsOids', iOid, 'acknowledged'], true, state);
    }
    case types.WS_VIEW_OBA_ALARM_ACK_FAILURE: {
      const { viewId, ackId, oid, error } = action.payload;
      if (!state[viewId].ackStatus[ackId]) {
        return state;
      }
      const { alarmsOids } = state[viewId].ackStatus[ackId];
      const iOid = _.findIndex(_.propEq('oid', oid), alarmsOids);
      return _.set([viewId, 'ackStatus', ackId, 'alarmsOids', iOid, 'ackError'], String(error), state);
    }
    case types.WS_VIEW_ALARM_COLLAPSE: {
      const { viewId, oid } = action.payload;
      const collapseAlarm = _.set([viewId, 'lines', oid, 'collapsed'], true);
      return collapseAlarm(state);
    }
    case types.WS_VIEW_ALARM_UNCOLLAPSE: {
      const { viewId, oid } = action.payload;
      const uncollapseAlarm = _.set([viewId, 'lines', oid, 'collapsed'], false);
      return uncollapseAlarm(state);
    }
    case types.INJECT_DATA_RANGE: {
      const {
        dataToInject,
        newViewMap,
        newExpectedRangeIntervals,
      } = action.payload;
      if (_.isEmpty(dataToInject)) {
        return state;
      }

      // since now, state will change
      let newState = state;
      const viewIds = Object.keys(state);
      for (let i = 0; i < viewIds.length; i += 1) {
        const viewId = viewIds[i];
        // Data Selection
        const epSubState =
          selectDataPerView(newViewMap[viewId], newExpectedRangeIntervals, dataToInject);
        if (Object.keys(epSubState).length !== 0) {
          // Data injection
          const viewState = viewRangeAdd(
            newState[viewId],
            viewId,
            epSubState
          );
          if (viewState !== newState[viewId]) {
            newState = { ...newState, [viewId]: viewState };
          }
        }
      }
      return newState || {};
    }
    case types.WS_VIEWDATA_CLEAN: {
      const { previousDataMap, dataMap } = action.payload;
      // since now, state will change
      let newState = state;
      const viewIds = Object.keys(state);
      for (let i = 0; i < viewIds.length; i += 1) {
        const viewId = viewIds[i];
        const viewData = state[viewId];

        // Cleaning
        const subState = cleanCurrentViewData(
          viewData,
          previousDataMap.perView[viewId],
          dataMap.perView[viewId],
          previousDataMap.expectedRangeIntervals,
          dataMap.expectedRangeIntervals
        );
        if (subState !== viewData) {
          newState = { ...newState, [viewId]: subState || initialSubState };
        }
      }
      return newState;
    }
    default:
      return state;
  }
}

export const getOnboardAlarmViewData = state => state.OnboardAlarmViewData;

export const getData = (state, { viewId }) => state.OnboardAlarmViewData[viewId];

export const getAckStatus = createSelector(
  getData,
  (state, { ackId }) => ackId,
  (data, ackId) => _.get(['ackStatus', ackId], data)
);

export const getDataLines = createSelector(
  getData,
  data => _.flatMap((lineId) => {
    const alarm = data.lines[lineId];
    const alarmWithoutParameters = _.omit('parameters', alarm);
    const parameters = _.isEmpty(alarm.parameters) || alarm.collapsed ? [] : [
      {
        type: 'parameter_header',
        alarm: alarmWithoutParameters,
      },
      ...alarm.parameters.map(parameter => ({
        type: 'parameter',
        data: parameter,
        alarm: alarmWithoutParameters,
      })),
    ];
    return [
      {
        type: 'alarm',
        data: alarmWithoutParameters,
        alarm: alarmWithoutParameters,
      },
      ...parameters,
    ];
  }, data.indexes)
);
