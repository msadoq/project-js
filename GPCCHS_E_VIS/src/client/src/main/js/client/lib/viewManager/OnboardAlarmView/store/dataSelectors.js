import { createSelector } from 'reselect';
import _ from 'lodash/fp';
import moment from 'moment';
import _last from 'lodash/last';
import _get from 'lodash/get';

import { getOnboardAlarmViewData, getData } from './dataReducer';
import * as constants from '../../../constants';
import { getViewTitle } from '../../../store/reducers/views';

const MODES_WORDING = {
  [constants.ALARM_MODE_ALL]: 'All',
  [constants.ALARM_MODE_TOACKNOWLEDGE]: 'To Acknowledge',
};

const getEntryPointsByViewId = (state, { viewId }) => (
  _.get(['OnboardAlarmViewConfiguration', viewId, 'entryPoints'], state)
);

const getFullTitle = createSelector(
  getViewTitle,
  getEntryPointsByViewId,
  (title, [entryPoint]) => (
    `${title} (${MODES_WORDING[entryPoint.connectedData.mode]})`
  )
);

const getCount = createSelector(
  getOnboardAlarmViewData,
  (dataState) => {
    if (!dataState || !Object.keys(dataState).length) {
      return { all: 0 };
    }
    const viewIds = Object.keys(dataState || {});
    const count = {};
    let countAll = 0;
    // Loop on views
    for (let i = 0; i < viewIds.length; i += 1) {
      const viewId = viewIds[i];
      const viewState = dataState[viewId];
      // lines is a table with all alarms
      const ptNb = viewState.lines.length;
      count[viewId] = ptNb;
      countAll += ptNb;
    }
    // Add all
    count.all = countAll;
    return count;
  }
);


const getLastValue = createSelector(
  (state, { viewId }) => getData(state, { viewId }),
  (viewData) => {
    if (!viewData) {
      return null;
    }
    // TODO adapt value to viewData
    const lastTimestamp = _last(_get(viewData, ['indexes']));
    const lastValue = _last(_get(viewData, ['lines']));
    if (!lastTimestamp || !lastValue) {
      return null;
    }
    const timestamp = moment(lastTimestamp).utc().toISOString();
    const symbol = _get(lastValue, 'symbol');
    const value = (symbol !== undefined && symbol !== null) ? symbol : _get(lastValue, 'value');
    return { timestamp, value };
  }
);

export default {
  getFullTitle,
  getEntryPointsByViewId,
  getCount,
  getLastValue,
};
