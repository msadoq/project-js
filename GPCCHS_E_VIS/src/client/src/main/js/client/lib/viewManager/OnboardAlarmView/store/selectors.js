import _ from 'lodash/fp';
import { createSelector } from 'reselect';
import { getData } from './dataReducer';
import { getSortMode, getSortColumn } from '../../GroundAlarmView/store/uiReducer';
import sortDataBy from '../../commonData/sortDataBy';

export const getSortedIndexes = createSelector(
  getData,
  getSortMode,
  getSortColumn,
  ({ lines, indexes }, sortMode, column) => (
    sortDataBy(oid => lines[oid].rawAlarm[column], sortMode, indexes)
  )
);

export const getDataRows = createSelector(
  getData,
  getSortedIndexes,
  (data, sortedIndexes) => _.flatMap((lineId) => {
    const alarm = data.lines[lineId];
    const alarmWithoutParameters = _.omit('parameters', alarm);
    const parameters = _.isEmpty(alarm.parameters) || alarm.collapsed ? [] : [
      {
        type: 'parameter_header',
        alarm: alarmWithoutParameters,
      },
      {
        type: 'parameter_header_title',
        alarm: alarmWithoutParameters,
      },
      ...alarm.parameters.map((parameter, index) => ({
        type: 'parameter',
        data: parameter,
        alarm: alarmWithoutParameters,
        parameterIndex: index,
      })),
    ];
    return [
      {
        type: 'alarm',
        data: alarmWithoutParameters,
        alarm: alarmWithoutParameters,
      },
      ...parameters,
    ];
  }, sortedIndexes)
);
