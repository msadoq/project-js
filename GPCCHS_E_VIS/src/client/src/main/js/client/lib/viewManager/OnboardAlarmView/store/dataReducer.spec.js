import { freezeArgs } from 'common/jest';
import * as types from 'store/types';
import * as constants from 'viewManager/constants';
import dataReducer, { getData } from './dataReducer';

import injectDataRangeAction from '../../GroundAlarmView/store/mocks/injectDataRangeAction.json';
import cleanViewDataAction from '../../GroundAlarmView/store/mocks/cleanViewDataAction.json';

const OBA = constants.VM_VIEW_ONBOARDALARM;
const PLOT = constants.VM_VIEW_PLOT;

const reducer = freezeArgs(dataReducer);

describe('OnboardAlarmView:dataReducer', () => {
  describe('reducer', () => {
    test('unknown action does nothing', () => {
      expect(reducer({}, { type: 'DUMMY_ACTION' })).toEqual({});
    });
    describe('documents', () => {
      test('add a view, open a view or reload a view should set initial sub state for the view', () => {
        const initialSubState = { v1: { indexes: [], lines: {} } };
        const state = { v1: { data: true } };

        expect(
          reducer(state, { type: types.WS_VIEW_RELOAD, payload: { view: { type: PLOT, uuid: 'v1' } } })
        ).toBe(state);
        expect(
          reducer(state, { type: types.WS_VIEW_RELOAD, payload: { view: { type: OBA, uuid: 'v1' } } })
        ).toEqual(initialSubState);
        expect(
          reducer(state, { type: types.WS_VIEW_ADD_BLANK, payload: { view: { type: OBA, uuid: 'v1' } } })
        ).toEqual(initialSubState);
        expect(
          reducer(state, { type: types.WS_VIEW_OPENED, payload: { view: { type: OBA, uuid: 'v1' } } })
        ).toEqual(initialSubState);
      });
      test('open a page or a workspace should set initial sub state for each related views', () => {
        const state = { v1: { data: true } };
        const payload = {
          views: [
            { type: PLOT, uuid: 'v0' },
            { type: OBA, uuid: 'v1' },
            { type: OBA, uuid: 'v2' },
            { type: OBA, uuid: 'v3' },
          ],
        };
        const finalState = {
          v1: { lines: {}, indexes: [] },
          v2: { lines: {}, indexes: [] },
          v3: { lines: {}, indexes: [] },
        };
        expect(reducer(state, { type: types.WS_PAGE_OPENED, payload })).toEqual(finalState);
        expect(reducer(state, { type: types.WS_WORKSPACE_OPENED, payload })).toEqual(finalState);
      });
      test('clean state', () => {
        expect(reducer({ a: true }, { type: types.DATA_REMOVE_ALL_VIEWDATA })).toEqual({});
      });
      test('clean state on closing workspace', () => {
        expect(reducer({ a: true }, { type: types.HSC_CLOSE_WORKSPACE })).toEqual({});
      });
      test('clean sub state when close a view', () => {
        const state = { v1: { data: true }, v2: { data: true } };
        const payload = { viewId: 'v1' };
        expect(
          reducer(state, { type: types.WS_VIEW_CLOSE, payload })
        ).toEqual({ v2: { data: true } });
      });
      test('clean all related views sub state when close a page', () => {
        const state = { v1: { data: true }, v2: { data: true }, v3: { data: true } };
        const payload = { viewIds: ['v1', 'v3', 'unknown_view_id'] };
        expect(
          reducer(state, { type: types.WS_PAGE_CLOSE, payload })
        ).toEqual({ v2: { data: true } });
      });
    });

    describe('data consumption', () => {
      const initialState = {
        v1: { lines: {}, indexes: [] },
        v2: { lines: {}, indexes: [] },
        v3: { lines: {}, indexes: [] },
      };
      const state = reducer(initialState, injectDataRangeAction);
      test('inject onboardAlarm data', () => {
        expect(state).toMatchSnapshot();
      });
      test('clean view data', () => {
        expect(reducer(state, cleanViewDataAction)).toEqual({
          v1: { lines: {}, indexes: [] },
          v2: { lines: {}, indexes: [] },
          v3: { lines: {}, indexes: [] },
        });
      });
    });
  });
  describe('selectors', () => {
    describe('getData', () => {
      const state = {
        OnboardAlarmViewData: {
          v3: {
            stuff: true,
          },
        },
      };
      test('get view data', () => {
        expect(getData(state, { viewId: 'v3' })).toEqual({ stuff: true });
      });
      test('get undefined when view is unknown', () => {
        expect(getData(state, { viewId: 'unknown_view' })).toBe(undefined);
      });
    });
  });
});
