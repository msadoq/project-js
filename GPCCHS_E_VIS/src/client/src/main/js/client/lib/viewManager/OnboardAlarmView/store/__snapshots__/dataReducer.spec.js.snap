// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardAlarmView:dataReducer getData getData should flatten alarms and transitions (in "lines" field) 1`] = `
Object {
  "stuff": true,
}
`;
